tests/hierarchy_driver.cc

```diff
@@ -63,33 +64,48 @@ void matrix_free_two_grids(std::shared_ptr<boost::property_tree::ptree> params)
       std::make_shared<TestMFMeshEvaluator<dim, fe_degree, double>>(
           mf_laplace._dof_handler, mf_laplace._constraints,
           mf_laplace._laplace_operator, material_property);
-  mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params, timer);
-
-  // We want to do 20 V-cycle iterations. The rhs of is zero.
-  // Use D(istributed)Vector because deal has its own Vector class
-  DVector residual(rhs);
-  unsigned int const n_cycles = 20;
-  std::vector<double> res(n_cycles + 1);
 
-  mf_laplace._laplace_operator.vmult(residual, solution);
-  residual.sadd(-1., 1., rhs);
-  auto const residual0_norm = residual.l2_norm();
+  mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params, timer);
 
-  std::cout << std::scientific;
-  res[0] = 1.0;
-  for (unsigned int i = 0; i < n_cycles; ++i)
+  if (!test_preconditioner)
   {
-    hierarchy.vmult(solution, rhs);
+    // We want to do 20 V-cycle iterations. The rhs of is zero.
+    // Use D(istributed)Vector because deal has its own Vector class
+    DVector residual(rhs);
+    unsigned int const n_cycles = 20;
+    std::vector<double> res(n_cycles + 1);
 
     mf_laplace._laplace_operator.vmult(residual, solution);
     residual.sadd(-1., 1., rhs);
-    double rel_residual = residual.l2_norm() / residual0_norm;
-    res[i + 1] = rel_residual;
-  }
+    auto const residual0_norm = residual.l2_norm();
 
-  double const conv_rate = res[n_cycles] / res[n_cycles - 1];
-  pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
-        << conv_rate << std::endl;
+    std::cout << std::scientific;
+    res[0] = 1.0;
+    for (unsigned int i = 0; i < n_cycles; ++i)
+    {
+      hierarchy.vmult(solution, rhs);
+
+      mf_laplace._laplace_operator.vmult(residual, solution);
+      residual.sadd(-1., 1., rhs);
+      double rel_residual = residual.l2_norm() / residual0_norm;
+      res[i + 1] = rel_residual;
+    }
+
+    double const conv_rate = res[n_cycles] / res[n_cycles - 1];
+    pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
+          << conv_rate << std::endl;
+  }
+  else
+  {
+    dealii::SolverControl solver_control(solution.size(), 1.e-6, true, true);
+    dealii::SolverCG<DVector> solver(solver_control);
+
+    solver.solve(mf_laplace._laplace_operator, solution, rhs,
+                 //                 dealii::PreconditionIdentity()
+                 hierarchy);
```

Comment:
Did you mean to leave that commented line?

Reply:
Yes, that was on purpose. I see this test as the one we want to profile with. Hence, I like to have a reminder for what I need to change if I don't want a preconditioner.

Anyway, I can remove it if you like that better.

Reply:
OK to leave if you add a comment above